### PR TITLE
[feat] KMeans implementation

### LINKED ISSUE #1714: Implement kmeans
## 🚀 Feature

Add a `kornia.contrib.kmeans` functionality that clusters batches of n-dimensional sets of features.

Reference: https://github.com/subhadarship/kmeans_pytorch/

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by avinashselvam**
#### Changes
`kornia/contrib/kmeans.py` implements kmeans using pytorch. runs on both cpu and gpu. 
No extra dependencies

New PR created from `main` replacing https://github.com/kornia/kornia/pull/2304 cause that was created from `master`

Fixes #1714

#### Type of change

- [x] 🧪 Tests Cases
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by avinashselvam** [comment]
Hi I was able to fix the errors in docstests. I'm unsure how to fix the failures in typing. I'm initialising 2 members as `None` and later setting them to tensors in `fit()`. I'm not used to typing in python. I tried adding `:Tensor` and `:Tensor|None` and neither of em work. Would like some help here. Thanks

File: kornia/contrib/kmeans.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+        # X = X.to(self.device)
+        current_centers = self.cluster_centers
+
+        previous_centers: Tensor | None = None
+        iteration: int = 0
+
+        while True:
+            # find distance between X and current_centers
+            distance: Tensor = self._pairwise_euclidean_distance(X, current_centers)
```
Comment (edgarriba) [inline kornia/contrib/kmeans.py:124]: you can use existing `euclidean_distance`: https://github.com/kornia/kornia/blob/69a697020a5f63ba7d46369673709c4f23d72313/kornia/geometry/linalg.py#L241

File: kornia/contrib/kmeans.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                selected = torch.nonzero(cluster_assignment == index).squeeze()
+                selected = torch.index_select(X, 0, selected)
+                # edge case when a certain cluster centre has no points assigned to it
+                # just choose a random point as it's update
+                if selected.shape[0] == 0:
+                    selected = X[torch.randint(len(X), (1,))]
+                current_centers[index] = selected.mean(dim=0)
+
+            # sum of distance of how much the newly computed clusters have moved from their previous positions
+            center_shift = torch.sum(torch.sqrt(torch.sum((current_centers - previous_centers) ** 2, dim=1)))
```
Comment (edgarriba) [inline kornia/contrib/kmeans.py:140]: probably here we can reuse the euclidean distance routine if we adjust the function to pass the reduction axis ? @johnnv1

File: kornia/contrib/kmeans.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        # x and cluster_centers should have same number of columns
+        KORNIA_CHECK(
+            x.shape[1] == self.final_cluster_centers.shape[1],
+            f"Dimensions at position 1 of x and cluster_centers do not match. \
+                {x.shape[1]} != {self.final_cluster_centers.shape[1]}",
+        )
+
+        distance = self._pairwise_euclidean_distance(x, self.final_cluster_centers)
+        cluster_assignment = torch.argmin(distance, axis=1)
+        return cluster_assignment
```
Comment (edgarriba) [inline kornia/contrib/kmeans.py:172]:
```suggestion
        return distance.argmin(-1)
```

Comment (edgarriba) [reply kornia/contrib/kmeans.py:172]: to avoid torch imports

File: kornia/contrib/kmeans.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+        self.final_cluster_assignments = None
+        self.final_cluster_centers = None
+
+        if seed is not None:
+            torch.manual_seed(seed)
+
+    def get_cluster_centers(self) -> Tensor:
+        KORNIA_CHECK(self.final_cluster_centers is not None, "Model has not been fit to a dataset")
+        return self.final_cluster_centers
```
Comment (edgarriba) [inline kornia/contrib/kmeans.py:55]: naming a bit confusing respect to the internal members `self.cluster_centers`. Maybe rename `self.current_cluster_centers`  or similar ?

File: kornia/contrib/kmeans.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        current_centers = self.cluster_centers
+
+        previous_centers: Tensor | None = None
+        iteration: int = 0
+
+        while True:
+            # find distance between X and current_centers
+            distance: Tensor = self._pairwise_euclidean_distance(X, current_centers)
+
+            cluster_assignment = torch.argmin(distance, dim=1)
```
Comment (edgarriba) [inline kornia/contrib/kmeans.py:126]:
```suggestion
            cluster_assignment = distances.argmin(1)
```

File: tests/contrib/test_kmeans.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        order = torch.argsort(centers[:, 0]).tolist()
+        new_classes = {old_class: new_class for new_class, old_class in enumerate(order)}
+
+        ordered_centers = centers[order]
+        oredered_prediciton = [new_classes[predicted_class] for predicted_class in prediciton]
+
+        self.assert_close(ordered_centers, expected_centers, atol=2, rtol=0.1)
+        assert oredered_prediciton == expected_prediciton
+
+    def test_gradcheck(self, device):
```
Comment (edgarriba) [inline tests/contrib/test_kmeans.py:106]: not sure we need this one here ? worth to add though `torchscript/dynamo` tests

File: kornia/contrib/kmeans.py (hunk 1 above)
Comment (avinashselvam) [reply kornia/contrib/kmeans.py:124]: Hi line 254 does `KORNIA_CHECK(x.shape == y.shape)` which prevents me from using this function. As `len(X) != len(current_centers)`

File: tests/contrib/test_kmeans.py (hunk 6 above)
Comment (avinashselvam) [reply tests/contrib/test_kmeans.py:106]: I'm not sure about what dynamo is. Where can I read about it? From other test cases it seems like we make an optimised version of the op and test the module again.

Comment (johnnv1) [reply tests/contrib/test_kmeans.py:106]: Hi @avinashselvam, dynamo is the [`torch.compile` API](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html)

Here is an example of how we are doing in our tests

https://github.com/kornia/kornia/blob/e40da282ba84b02ed8f3958dc220ce76d20572aa/tests/color/test_gray.py#L91

File: kornia/contrib/kmeans.py (hunk 2 above)
Comment (johnnv1) [reply kornia/contrib/kmeans.py:140]: worth a try

**Review by johnnv1 — COMMENTED** [review]
This solves the typing issue:

```diff
diff --git a/kornia/contrib/kmeans.py b/kornia/contrib/kmeans.py
index a381a655..369b5cbe 100644
--- a/kornia/contrib/kmeans.py
+++ b/kornia/contrib/kmeans.py
@@ -28,7 +28,7 @@ class KMeans:
     def __init__(
         self,
         num_clusters: int,
-        cluster_centers: Tensor,
+        cluster_centers: Tensor | None,
         tolerance: float = 10e-4,
         max_iterations: int = 0,
         seed: int | None = None,
@@ -40,23 +40,32 @@ class KMeans:
             KORNIA_CHECK_SHAPE(cluster_centers, ["C", "D"])
 
         self.num_clusters = num_clusters
-        self.cluster_centers = cluster_centers
+        self._cluster_centers = cluster_centers
         self.tolerance = tolerance
         self.max_iterations = max_iterations
 
-        self.final_cluster_assignments = None
-        self.final_cluster_centers = None
+        self._final_cluster_assignments: None | Tensor = None
+        self._final_cluster_centers: None | Tensor = None
 
         if seed is not None:
             torch.manual_seed(seed)
 
-    def get_cluster_centers(self) -> Tensor:
-        KORNIA_CHECK(self.final_cluster_centers is not None, "Model has not been fit to a dataset")
-        return self.final_cluster_centers
+    @property
+    def cluster_centers(self) -> Tensor:
+        if isinstance(self._final_cluster_centers, Tensor):
+            return self._final_cluster_centers
+        elif isinstance(self._cluster_centers, Tensor):
+            return self._cluster_centers
 
-    def get_cluster_assignments(self) -> Tensor:
-        KORNIA_CHECK(self.final_cluster_assignments is not None, "Model has not been fit to a dataset")
-        return self.final_cluster_assignments
+        else:
+            raise TypeError("Model has not been fit to a dataset")
+
+    @property
+    def cluster_assignments(self) -> Tensor:
+        if isinstance(self._final_cluster_assignments, Tensor):
+            return self._final_cluster_assignments
+        else:
+            raise TypeError("Model has not been fit to a dataset")
 
     def _initialise_cluster_centers(self, X: Tensor, num_clusters: int) -> Tensor:
         """Chooses num_cluster points from X as the initial cluster centers.
@@ -103,18 +112,18 @@ class KMeans:
         # X should have only 2 dimensions
         KORNIA_CHECK_SHAPE(X, ["N", "D"])
 
-        if self.cluster_centers is None:
-            self.cluster_centers = self._initialise_cluster_centers(X, self.num_clusters)
+        if self._cluster_centers is None:
+            self._cluster_centers = self._initialise_cluster_centers(X, self.num_clusters)
         else:
             # X and cluster_centers should have same number of columns
             KORNIA_CHECK(
-                X.shape[1] == self.cluster_centers.shape[1],
+                X.shape[1] == self._cluster_centers.shape[1],
                 f"Dimensions at position 1 of X and cluster_centers do not match. \
-                {X.shape[1]} != {self.cluster_centers.shape[1]}",
+                {X.shape[1]} != {self._cluster_centers.shape[1]}",
             )
 
         # X = X.to(self.device)
-        current_centers = self.cluster_centers
+        current_centers = self._cluster_centers
 
         previous_centers: Tensor | None = None
         iteration: int = 0
@@ -123,7 +132,7 @@ class KMeans:
             # find distance between X and current_centers
             distance: Tensor = self._pairwise_euclidean_distance(X, current_centers)
 
-            cluster_assignment = torch.argmin(distance, dim=1)
+            cluster_assignment = distance.argmin(dim=1)
 
             previous_centers = current_centers.clone()
 
@@ -147,8 +156,8 @@ class KMeans:
             if self.max_iterations != 0 and iteration >= self.max_iterations:
                 break
 
-        self.final_cluster_assignments = cluster_assignment
-        self.final_cluster_centers = current_centers
+        self._final_cluster_assignments = cluster_assignment
+        self._final_cluster_centers = current_centers
 
     def predict(self, x: Tensor) -> Tensor:
         """Find the cluster center closest to each point in x.
@@ -162,11 +171,11 @@ class KMeans:
 
         # x and cluster_centers should have same number of columns
         KORNIA_CHECK(
-            x.shape[1] == self.final_cluster_centers.shape[1],
+            x.shape[1] == self.cluster_centers.shape[1],
             f"Dimensions at position 1 of x and cluster_centers do not match. \
-                {x.shape[1]} != {self.final_cluster_centers.shape[1]}",
+                {x.shape[1]} != {self.cluster_centers.shape[1]}",
         )
 
-        distance = self._pairwise_euclidean_distance(x, self.final_cluster_centers)
-        cluster_assignment = torch.argmin(distance, axis=1)
+        distance = self._pairwise_euclidean_distance(x, self.cluster_centers)
+        cluster_assignment = distance.argmin(dim=1)
         return cluster_assignment
```

I'll try to comment some points

File: kornia/contrib/kmeans.py (hunk 4 above)
Comment (johnnv1) [reply kornia/contrib/kmeans.py:55]: \+ 1 to just have one `cluster_center` "public" for the user's access, otherwise will be too much information...

In the type correction I commented on the PR, I suggest moving to "internal"/"private" members, and use `cluster_centers` and `cluster_assignments` as properties

File: kornia/contrib/kmeans.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    Example:
+    >>> kmeans = kornia.contrib.KMeans(3, None, 10e-4, 100, device, 0)
+    >>> kmeans.fit(torch.rand((1000, 5))
+    >>> predictions = kmeans.predict(torch.rand((10, 5)))
+    """
+
+    def __init__(
+        self,
+        num_clusters: int,
+        cluster_centers: Tensor,
```
Comment (johnnv1) [inline kornia/contrib/kmeans.py:31]: from the code, and your doc, it can also be `None` -- which will invoke `_initialise_cluster_centers`
```suggestion
        cluster_centers: Tensor | None,
```

File: kornia/contrib/kmeans.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        if cluster_centers is not None:
+            KORNIA_CHECK_SHAPE(cluster_centers, ["C", "D"])
+
+        self.num_clusters = num_clusters
+        self.cluster_centers = cluster_centers
+        self.tolerance = tolerance
+        self.max_iterations = max_iterations
+
+        self.final_cluster_assignments = None
+        self.final_cluster_centers = None
```
Comment (johnnv1) [inline kornia/contrib/kmeans.py:48]: If you need to initialize this variable, then you need to add a typing here, otherwise, the type checker will be mad because when it are updated to Tensor.

```suggestion
        self.final_cluster_assignments: None | Tensor = None
        self.final_cluster_centers: None | Tensor = None
```

Comment (johnnv1) [reply kornia/contrib/kmeans.py:48]: also, i advocate to maybe go with internal/private member for these

File: kornia/contrib/kmeans.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+        if seed is not None:
+            torch.manual_seed(seed)
+
+    def get_cluster_centers(self) -> Tensor:
+        KORNIA_CHECK(self.final_cluster_centers is not None, "Model has not been fit to a dataset")
```
Comment (johnnv1) [inline kornia/contrib/kmeans.py:54]: KORNIA_CHECK API, sadly didn't guard type right (even the KORNIA_CHECK_TYPE -- check https://github.com/kornia/kornia/blob/e40da282ba84b02ed8f3958dc220ce76d20572aa/kornia/core/check.py#L122), so to type checking didn't be mad, we need to manually type check it with `isinstance` else raise

File: tests/contrib/test_kmeans.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
+import pytest
+import torch
+
+import kornia
+from kornia.testing import BaseTester
```
Comment (johnnv1) [inline tests/contrib/test_kmeans.py:5]:
```suggestion
from testing.base import BaseTester
```

Then you can remove the unnecessary tests, like `test_gradcheck` -- it isn't a abstract class anymore 😄

File: kornia/contrib/kmeans.py (hunk 1 above)
Comment (edgarriba) [reply kornia/contrib/kmeans.py:124]: i think we can safely remove that check /cc @johnnv1

**Comment by avinashselvam** [comment]
- Getting `torch.compile` not supported in windows error.
 - `test_dynamo` is getting skipped on both my windows machine (with gpu) and mac m1.
 - However if I use `@torch.compile` on `fit` and `predict` functions `test_module` passes on m1.

How do I go about this?
My bad about the obvious bug in the `test_dynamo`

File: tests/contrib/test_kmeans.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        prediciton = kmeans.predict(predict_param, dtype=dtype, device=device).tolist()
+
+        kmeans_op = kornia.contrib.KMeans(*kmeans_params)
+        kmeans_op.fit = torch_optimizer(kmeans_op.fit)
+        kmeans_op.predict = torch_optimizer(kmeans_op.predict)
+
+        kmeans_op.fit(x)
+
+        centers_op = kmeans_op.cluster_centers
+        prediciton_op = kmeans_op.predict(predict_param, dtype=dtype, device=device).tolist()
```
Comment (johnnv1) [inline tests/contrib/test_kmeans.py:130]:
```suggestion
        prediciton = kmeans.predict(predict_param)

        kmeans_op = kornia.contrib.KMeans(*kmeans_params)
        kmeans_op.fit = torch_optimizer(kmeans_op.fit)
        kmeans_op.predict = torch_optimizer(kmeans_op.predict).detach().cpu().tolist()

        kmeans_op.fit(x)

        centers_op = kmeans_op.cluster_centers
        prediciton_op = kmeans_op.predict(predict_param).detach().cpu().tolist()
```

**Comment by johnnv1** [comment]
> How do I go about this?

no problem, we can create an issue to add it later... you can just add a `@pytest.mark.skip` marker for now 😄

File: kornia/contrib/kmeans.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+from __future__ import annotations
+
+import torch
+
+from kornia.core import Tensor
+from kornia.core.check import KORNIA_CHECK, KORNIA_CHECK_SHAPE
+
+
+class KMeans:
+    """Implements the kmeans clustering algorithm with euclidean distance as similarity measure.
```
Comment (johnnv1) [inline kornia/contrib/kmeans.py:12]: One last thing, can you add an entry on `docs/source/contrib.rs`t for it?

File: tests/contrib/test_kmeans.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        kmeans_op.fit = torch_optimizer(kmeans_op.fit)
+        kmeans_op.predict = torch_optimizer(kmeans_op.predict).detach().cpu().tolist()
+
+        kmeans_op.fit(x)
+
+        centers_op = kmeans_op.cluster_centers
+        prediciton_op = kmeans_op.predict(predict_param).detach().cpu().tolist()
+
+        self.assert_close(centers, centers_op)
+        assert prediciton == prediciton_op
```
Comment (johnnv1) [inline tests/contrib/test_kmeans.py:133]:
```suggestion
        prediciton = kmeans.predict(predict_param)

        kmeans_op = kornia.contrib.KMeans(*kmeans_params)
        kmeans_op.fit = torch_optimizer(kmeans_op.fit)
        kmeans_op.predict = torch_optimizer(kmeans_op.predict)

        kmeans_op.fit(x)

        centers_op = kmeans_op.cluster_centers
        prediciton_op = kmeans_op.predict(predict_param)

        self.assert_close(centers, centers_op)
        self.assert_close(prediciton, prediciton_op)
```

**Review by johnnv1 — APPROVED** [review]

**Comment by edgarriba** [comment]
@avinashselvam could you contribute at some point with a small tutorial? Something funny e.g to cluster some features from a vision network?

**Comment by avinashselvam** [comment]
Hi yes for sure. Need a week or two as it's kinda busy rn. I'll get to it after that.
